src/upgrade.py

```diff
@@ -0,0 +1,281 @@
+            if (
+                # TODO charm versioning: Un-comment when charm versioning specification is
+                # implemented. Charm versions with git hash (temporary implementation) cannot be
+                # compared with `<`
+                # previous_versions["charm"] > current_versions["charm"] or
+                previous_versions["charm"].major
+                != current_versions["charm"].major
+            ):
+                logger.debug(
+                    f'{previous_versions["charm"]=} incompatible with {current_versions["charm"]=}'
+                )
+                return False
+            if (
+                previous_versions["workload"] > current_versions["workload"]
+                or previous_versions["workload"].major != current_versions["workload"].major
+                or previous_versions["workload"].minor != current_versions["workload"].minor
+            ):
+                logger.debug(
+                    f'{previous_versions["workload"]=} incompatible with {current_versions["workload"]=}'
+                )
+                return False
+            logger.debug(
+                f"Versions before upgrade compatible with versions after upgrade {previous_version_strs=} {self._current_versions=}"
+            )
+            return True
+        except KeyError as exception:
+            logger.debug(f"Version missing from {previous_versions=}", exc_info=exception)
+            return False
+
+    @property
+    def in_progress(self) -> bool:
+        logger.debug(f"{self._app_workload_version=} {self._unit_workload_versions=}")
+        return any(
+            version != self._app_workload_version
+            for version in self._unit_workload_versions.values()
+        )
+
+    @property
+    def _sorted_units(self) -> list[ops.Unit]:
+        """Units sorted from highest to lowest unit number"""
+        return sorted((self._unit, *self._peer_relation.units), key=_unit_number, reverse=True)
+
+    @property
+    def _unit_active_status(self) -> ops.ActiveStatus:
+        """Status shown during upgrade if unit is healthy"""
+        return ops.ActiveStatus(self._current_versions["charm"])
+
+    @property
+    def unit_juju_status(self) -> typing.Optional[ops.StatusBase]:
+        if self.in_progress:
+            return self._unit_active_status
+
+    @property
+    def app_status(self) -> typing.Optional[ops.StatusBase]:
+        if self.in_progress:
+            if len(self._sorted_units) >= 2 and self._partition > _unit_number(
+                self._sorted_units[1]
+            ):
+                # User confirmation needed to resume upgrade (i.e. upgrade second unit)
+                # Statuses over 120 characters are truncated in `juju status` as of juju 3.1.6 and
+                # 2.9.45
+                return ops.BlockedStatus(
+                    f"Upgrading. Verify highest unit is healthy & run `{RESUME_ACTION_NAME}` action. To rollback, `juju refresh` to last revision"
```

Comment:
Keep it for now AS IS. I like a shorter approach, but we need to release 8.0/stable and receive UX feedback... it will be improved in the future for sure.